### PR TITLE
Move to moe-kernels package and switch to common MoE layer

### PR DESCRIPTION
**Pull request by danieldk**
# What does this PR do?

This change introduces the new `moe-kernels` package:

- Add `moe-kernels` as a dependency.
- Introduce a `SparseMoELayer` module that can be used by MoE models.
- Port over Mixtral and Deepseek.

GPTQ-Marlin support and porting DBRX is for follow-up PRs.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by danieldk** [comment]
The CI failure looks unrelated.

File: server/text_generation_server/models/custom_modeling/flash_mixtral_modeling.py [hunk 1]
```diff
@@ -45,6 +43,7 @@
     SpeculativeHead,
     get_linear,
 )
+from text_generation_server.layers.moe import SparseMoELayer
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/flash_mixtral_modeling.py:46]: Can this be imported on ipex?

**Review by Narsil — COMMENTED** [review]
LGTM.

Have you checked th tests before the PR on mixtral see if the output is the same ?

**Comment by danieldk** [comment]
> Have you checked th tests before the PR on mixtral see if the output is the same ?

Yeah, I created the test first on the `main` branch.

File: integration-tests/models/test_flash_mixtral.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+
+@pytest.mark.skip(reason="requires > 4 shards")
+@pytest.mark.asyncio
+async def test_flash_mixtral(flash_mixtral, response_snapshot):
+    response = await flash_mixtral.generate(
+        "Test request", max_new_tokens=10, decoder_input_details=True
+    )
+
+    assert response == response_snapshot
```
[the remark this answers is not in the document]
Comment (danieldk) [reply integration-tests/models/test_flash_mixtral.py]: Fixed!

File: integration-tests/models/test_flash_mixtral.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+
+@pytest.mark.skip(reason="requires > 4 shards")
+@pytest.mark.asyncio
+async def test_flash_mixtral_load(flash_mixtral, generate_load, response_snapshot):
+    responses = await generate_load(
+        flash_mixtral, "Test request", max_new_tokens=10, n=4
+    )
+
+    assert len(responses) == 4
```
[the remark this answers is not in the document]
Comment (danieldk) [reply integration-tests/models/test_flash_mixtral.py]: Fixed.

File: server/text_generation_server/models/custom_modeling/flash_mixtral_modeling.py (hunk 1 above)
Comment (danieldk) [reply server/text_generation_server/models/custom_modeling/flash_mixtral_modeling.py:46]: I changed the import in `text_generation_server.layers.moe.unquantized` conditional on IPEX not being used (like the old code).

File: integration-tests/models/__snapshots__/test_flash_mixtral/test_flash_mixtral_all_params.json [hunk 4]
```diff
@@ -64,7 +64,7 @@
       },
       {
         "id": 369,
-        "logprob": -0.06585693,
+        "logprob": 0.0,
```
Comment (Narsil) [inline integration-tests/models/__snapshots__/test_flash_mixtral/test_flash_mixtral_all_params.json:67]: Huh ?

Comment (danieldk) [reply integration-tests/models/__snapshots__/test_flash_mixtral/test_flash_mixtral_all_params.json:67]: Guess that's what a temperature of 0.5 does though:

```
❯ git grep '"logprob": 0\.0' | uniq | head -n5
integration-tests/models/__snapshots__/test_bloom_560m/test_bloom_560m.json:        "logprob": 0.0,
integration-tests/models/__snapshots__/test_bloom_560m/test_bloom_560m_all_params.json:        "logprob": 0.0,
integration-tests/models/__snapshots__/test_bloom_560m_sharded/test_bloom_560m_sharded.json:        "logprob": 0.0,
integration-tests/models/__snapshots__/test_flash_awq/test_flash_llama_awq_all_params.json:        "logprob": 0.0,
integration-tests/models/__snapshots__/test_flash_deepseek_v2/test_flash_deepseek_v2_all_params.json:        "logprob": 0.0,
```

**Review by Narsil — APPROVED** [review]

**Pull request merged**